typing/datarepr.ml

```diff
@@ -21,20 +21,22 @@ open Types
 open Btype
 
 (* Simplified version of Ctype.free_vars *)
-let free_vars ?(param=false) ty =
+let free_vars ?(param=false) ?(include_row_variables=true) ty =
   let ret = ref TypeSet.empty in
   let rec loop ty =
     if try_mark_node ty then
       match get_desc ty with
       | Tvar _ ->
           ret := TypeSet.add ty !ret
       | Tvariant row ->
-          iter_row loop row;
-          if not (static_row row) then begin
-            match get_desc (row_more row) with
-            | Tvar _ when param -> ret := TypeSet.add ty !ret
-            | _ -> loop (row_more row)
-          end
+        if include_row_variables
+        then iter_row loop row
+        else iter_row_fields loop row;
+        if not (static_row row) && include_row_variables then begin
```

Comment:
This looks wrong to me. I think that fields can be spread across many `row_more`s, in a little linked list, even for closed rows.